backend/ethereum/subscription/resistanteventsub_test.go

```diff
@@ -0,0 +1,143 @@
+		require.NoError(err)
+		log.Debugf("Needed confirmations: %d", finality)
+		// Send and Confirm the TX. The simulated backend already mined a block here,
+		// so the TX has 1 confirmation now.
+		s.ConfirmTx(s.IncAllowance(ctx), true)
+		// Wait `finality-1` blocks.
+		for j := int32(0); j < finality-1; j++ {
+			NoEvent(require, sub)
+			log.Debug("Commit")
+			s.SB.Commit()
+		}
+		log.Debug("Waiting for confirm")
+		OneEvent(require, sub)
+		sub.Close()
+		s.SB.Commit() // Make sure the next EventSub does not get some old block.
+	}
+}
+
+// TestResistantEventSub_FinalityDepth tests that a TX is confirmed exactly
+// after `finalityDepth` blocks even when reorgs occur.
+func TestResistantEventSub_Reorg(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	rng := pkgtest.Prng(t)
+	require := require.New(t)
+	s := test.NewTokenSetup(ctx, t, rng)
+
+	for i := 0; i < 3; i++ {
+		finality := rng.Int31n(20) + 2
+		sub, err := subscription.Subscribe(ctx, s.CB, s.Contract, event, 0, uint64(finality))
+		require.NoError(err)
+
+		for i := 0; i < 20; i++ {
+			s.SB.Commit()
+		}
+		// Send and Confirm the TX
+		s.IncAllowance(ctx)
+		// Reorg until the block hight hits `finality` blocks.
+		for h := int64(0); h < int64(finality-1); {
+			d := rng.Int63n(10) + 1
+			l := rng.Int63n(10) + 1
+			NoEvent(require, sub)
+			log.Debugf("[h=%d] Reorg with depth: %d, length: %d", h, d, l)
+			s.SB.Reorg(ctx, uint64(d), func(txs []types.Transactions) []types.Transactions {
+				return append(txs, make([]types.Transactions, int(d+l)-len(txs))...)
+			})
+			h += l
+		}
+		log.Debug("done")
+		// Verify that the event arrived.
+		OneEvent(require, sub)
+		sub.Close()
+	}
+}
+
+// TestResistantEventSub_New checks that `NewResistantEventSub` panics for
+// `confirmations` < 1.
+func TestResistantEventSub_New(t *testing.T) {
+	require.PanicsWithValue(t, "confirmations needs to be at least 1", func() {
+		subscription.NewResistantEventSub(context.Background(), nil, nil, 0)
+	})
+}
+
+// NoEvent checks that no event can be read from `sub`.
+func NoEvent(require *require.Assertions, sub *subscription.ResistantEventSub) {
+	nEvents(require, 0, sub)
+}
+
+// OneEvent checks that exactly one event can be read from `sub`.
+func OneEvent(require *require.Assertions, sub *subscription.ResistantEventSub) {
+	nEvents(require, 1, sub)
+}
+
+// nEvents checks that exactly `n` events can be read from `sub`.
+func nEvents(require *require.Assertions, n int, sub *subscription.ResistantEventSub) {
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+	sink := make(chan *subscription.Event, n+1)
+
+	err := sub.Read(ctx, sink)
+	require.True(pctx.IsContextError(err))
+
+	for i := 0; i < n; i++ {
+		require.NotNil(<-sink)
+	}
+
+	select {
+	case event := <-sink:
+		require.Nil(event)
+	default:
+	}
+}
```

Comment:
What do you think about adding a test for the case that an event is removed by a reorg and should therefore not be emitted?

Reply:
Ok makes sense, analogous to the TX finality test.

Reply:
Wow… adding such a test fails, since indeed, the header subscription seems to be much faster than the event subscription.

Reply:
Looks like to increased that to a 1second wait.